File: claude_exp/xv7cpp/src/router/router_main.cpp
#include <cstdlib>
#include <iostream>
#include <memory>
#include <random>
#include <thread>

#include <nlohmann/json.hpp>

#include "router/router_config.h"
#include "router/router_session.h"
#include "router/upstream.h"
#include "shared/command_server.h"
#include "shared/log.h"
#include "shared/stats.h"
#include "shared/stop_event.h"

using json = nlohmann::json;
using namespace xv6::router;
using namespace xv6::shared;

namespace {

std::string parse_config_arg(int argc, char** argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (std::string(argv[i]) == "--config") {
            return argv[i + 1];
        }
    }
    throw std::runtime_error("missing --config <path>");
}

void send_json(httplib::Response& res, int status, const json& obj) {
    res.set_header("Content-Type", "application/json");
    res.set_content(obj.dump(), "application/json");
    res.status = status;
}

void wait_reestablish(StopEvent& stop_event, const RouterConfig& cfg) {
    static std::mt19937 gen(std::random_device{}());
    std::uniform_real_distribution<> dis(0.0, cfg.reconnect_jitter_seconds);

    int total_wait_ms = cfg.reestablish_seconds * 1000 + (int)(dis(gen) * 1000);
    int poll_interval_ms = 100;

    for (int elapsed = 0; elapsed < total_wait_ms; elapsed += poll_interval_ms) {
        if (stop_event.is_set()) break;
        stop_event.wait_for(std::chrono::milliseconds(poll_interval_ms));
    }
}

}  // namespace

int main(int argc, char** argv) {
    try {
        auto cfg_path = parse_config_arg(argc, argv);
        RouterConfig cfg = RouterConfig::from_file(cfg_path);
        StopEvent stop_event;
        Stats stats(cfg.yellow_threshold_seconds);

        Logger::instance().set_level(*parse_log_level(cfg.log_level));
        CommandServer cmd(cfg.command_port, stats, stop_event, cfg.command_bind_host,
                          cfg.command_auth_token);

        std::shared_ptr<Dispatcher> active_dispatcher;
        std::mutex active_dispatcher_mutex;
        cmd.register_route("/dispatcher/purge", {"POST"}, /*protected=*/true,
                          [&](const httplib::Request&, httplib::Response& res) {
                              std::lock_guard lock(active_dispatcher_mutex);
                              if (!active_dispatcher) {
                                  send_json(res, 503, {{"error", "no active session"}});
                                  return;
                              }
                              send_json(res, 200, active_dispatcher->purge());
                          });
        cmd.start();

        std::unique_ptr<UpstreamServer> srv_sock;
        if (cfg.upstream.mode == "server") {
            srv_sock = std::make_unique<UpstreamServer>(cfg.upstream);
        }

        while (!stop_event.is_set()) {
            std::unique_ptr<RouterSession> session;
            try {
                session = RouterSession::connect(cfg, stats, stop_event);
            } catch (const std::exception& e) {
                LOG_WARNING(std::string("failed to connect downstream: ") + e.what());
                wait_reestablish(stop_event, cfg);
                continue;
            }

            {
                std::lock_guard lock(active_dispatcher_mutex);
                active_dispatcher = std::shared_ptr<Dispatcher>(session->dispatcher(), [](Dispatcher*) {});
            }
            session->run_until_disconnect(srv_sock.get());
            {
                std::lock_guard lock(active_dispatcher_mutex);
                active_dispatcher = nullptr;
            }

            if (!stop_event.is_set()) {
                wait_reestablish(stop_event, cfg);
            }
        }

        cmd.stop();
        return 0;
    } catch (const std::exception& e) {
        LOG_ERROR(std::string("fatal: ") + e.what());
        return 1;
    }
}

File: claude_exp/routers/router_java/src/main/java/com/xv6/router/RouterMain.java
package com.xv6.router;

import com.xv6.shared.CommandServer;
import com.xv6.shared.LogLevels;
import com.xv6.shared.Stats;
import com.xv6.shared.StopEvent;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import java.util.logging.Logger;

/** Entry point, reconnect loop. Port of router_py's router/main.py. */
public final class RouterMain {

    private static final Logger logger = Logger.getLogger(RouterMain.class.getName());

    private RouterMain() {
    }

    public static void run(RouterConfig cfg, StopEvent stopEvent, Stats statsIn) throws Exception {
        Stats stats = statsIn != null ? statsIn : new Stats(cfg.yellowThresholdSeconds());

        // MUST come before CommandServer(...): CommandServer adds a LogBuffer handler to the
        // root logger, and setting the level after that still works, but doing it first mirrors
        // router_py's ordering requirement (there, a late basicConfig call is a no-op once the root
        // logger already has handlers - not an issue for java.util.logging, but kept in the same
        // order for parity).
        Logger.getLogger("").setLevel(LogLevels.parse(cfg.logLevel()));

        CommandServer cmd = new CommandServer(
                cfg.commandPort(), stats, stopEvent, cfg.commandBindHost(), cfg.commandAuthToken());

        AtomicReference<Dispatcher> activeDispatcher = new AtomicReference<>();

        cmd.register("/dispatcher/purge", List.of("POST"), true, exchange -> {
            Dispatcher dispatcher = activeDispatcher.get();
            if (dispatcher == null) {
                CommandServer.sendJson(exchange, 503, Map.of("error", "no active session"));
                return;
            }
            CommandServer.sendJson(exchange, 200, dispatcher.purge());
        });

        cmd.start();

        Upstream.UpstreamServer srvSock = null;
        if ("server".equals(cfg.upstream().mode())) {
            srvSock = new Upstream.UpstreamServer(cfg.upstream());
        }

        try {
            Random random = new Random();
            while (!stopEvent.isSet()) {
                RouterSession session;
                try {
                    session = RouterSession.connect(cfg, stats, stopEvent);
                } catch (IOException e) {
                    logger.warning("failed to connect downstream: " + e.getMessage());
                    waitReestablish(stopEvent, cfg, random);
                    continue;
                }

                activeDispatcher.set(session.dispatcher);
                session.runUntilDisconnect(srvSock);
                activeDispatcher.set(null);

                if (!stopEvent.isSet()) {
                    // Jitter avoids multiple routers sharing a downstream/crypto host
                    // reconnecting in lockstep.
                    waitReestablish(stopEvent, cfg, random);
                }
            }
        } finally {
            if (srvSock != null) {
                srvSock.close();
            }
        }
    }

    private static void waitReestablish(StopEvent stopEvent, RouterConfig cfg, Random random)
            throws InterruptedException {
        double jitter = random.nextDouble() * cfg.reconnectJitterSeconds();
        long millis = Math.round((cfg.reestablishSeconds() + jitter) * 1000);
        stopEvent.waitFor(millis, TimeUnit.MILLISECONDS);
    }

    public static void main(String[] args) throws Exception {
        String configPath = null;
        for (int i = 0; i < args.length; i++) {
            if ("--config".equals(args[i]) && i + 1 < args.length) {
                configPath = args[i + 1];
            }
        }
        if (configPath == null) {
            throw new IllegalArgumentException("usage: RouterMain --config <path>");
        }
        RouterConfig cfg = RouterConfig.fromFile(configPath);
        try {
            run(cfg, new StopEvent(), null);
            // com.sun.net.httpserver.HttpServer's internal dispatcher thread is not a daemon
            // thread (confirmed by testing: the process stayed alive after /stop with only that
            // thread left), so unlike router_py (where every thread genuinely is a daemon thread and
            // the interpreter exits on its own), this JVM needs an explicit exit once run()
            // returns. run() itself must stay exit-free, since RouterFullStackTest calls it
            // directly in-process and a System.exit() here would kill the test JVM too.
            System.exit(0);
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}
